vnext/JSI/Shared/ChakraRuntime.cpp

```diff
@@ -443,144 +471,94 @@ facebook::jsi::WeakObject ChakraRuntime::createWeakObject(const facebook::jsi::O
-
-  // hostFuncProxy keeps a reference to this Runtime, so we must keep this
-  // Runtime alive as long as hostFuncProxy is alive. We do so by making the
-  // garbage collector delete hostFuncProxy when funcRef is finalized. Since
-  // funcRef cannot out-live this Runtime, the reference stored in hostFuncProxy
-  // stays valid during its lifetime.
-  VerifyJsErrorElseThrow(JsSetObjectBeforeCollectCallback(
-      funcRef, hostFuncProxyWrapper.get(), [](JsRef /*ref*/, void *hostFuncProxyToDestroy) {
-        // We wrap hostFuncProxyToDestroy in a unique_ptr to avoid calling
-        // delete explicitly.
-        std::unique_ptr<HostFunctionProxy> wrapper{static_cast<HostFunctionProxy *>(hostFuncProxyToDestroy)};
-      }));
-
-  // We only call hostFuncProxyWrapper.release() after
-  // JsSetObjectBeforeCollectCallback succeeds. Otherwise, when
-  // JsSetObjectBeforeCollectCallback fails and an exception is thrown,
-  // the HostFunctionProxy that hostFuncProxyWrapper used to own will be leaked.
-  hostFuncProxyWrapper.release();
-
-  facebook::jsi::Object hostFuncObj = MakePointer<facebook::jsi::Object>(funcRef);
-
-  // We do not use the function
-  //   Object::setProperty(Runtime & runtime, const char *name, T &&value)
-  // here because it causes multiple copies of name.
-  setPropertyValue(
-      hostFuncObj,
-      createPropNameIDFromAscii(g_functionIsHostFunctionPropName, strlen(g_functionIsHostFunctionPropName)),
-      facebook::jsi::Value(true));
-
-  return hostFuncObj.getFunction(*this);
+  auto hostFunctionWrapper = std::make_unique<HostFunctionWrapper>(std::move(func), *this);
+  JsValueRef function = CreateExternalFunction(
+      GetJsRef(name), static_cast<int32_t>(paramCount), HostFunctionCall, hostFunctionWrapper.get());
+
+  auto hostFunctionHolder = CreateExternalObject(std::move(hostFunctionWrapper));
+  DefineProperty(
+      function,
+      m_propertyId.hostFunctionSymbol,
+      CreatePropertyDescriptor(hostFunctionHolder, PropertyAttibutes::DontEnumAndFrozen));
+
+  return MakePointer<facebook::jsi::Object>(function).getFunction(*this);
 }
 
 facebook::jsi::Value ChakraRuntime::call(
     const facebook::jsi::Function &func,
     const facebook::jsi::Value &jsThis,
     const facebook::jsi::Value *args,
     size_t count) {
-  // We must store these ChakraObjectRefs on the stack to make sure that they do
-  // not go out of scope when JsCallFunction is called.
-  ChakraObjectRef thisRef = ToChakraObjectRef(jsThis);
-  std::vector<ChakraObjectRef> argRefs = ToChakraObjectRefs(args, count);
-
-  std::vector<JsValueRef> argsWithThis = ConstructJsFunctionArguments(thisRef, argRefs);
-  assert(argsWithThis.size() <= (std::numeric_limits<unsigned short>::max)());
+  size_t jsArgCount = count + 1; // for 'this' argument.
+  ChakraVerifyElseThrow(jsArgCount <= MaxCallArgCount, "Argument count exceeds MaxCallArgCount");
```

Comment:
>MaxCallArgCount [](start = 38, length = 15)

Shouldn't this be a bit more dynamic? I.e. does JSI (or any egine for that) have a limit of 32 args?
Perhaps the code should be optimized for a few cases: 
< 8 args, => on stack size 8
<  32 args => on stack size 32
>= 32 args => on heap

Reply:
Great suggestion: I will see how to do it.

---
In reply to: [470106454](https://github.com/microsoft/react-native-windows/pull/5710#discussion_r470106454) [](ancestors = 470106454)

Reply:
I was able to specialize it for <= 8 and > 8 case

---
In reply to: [470186502](https://github.com/microsoft/react-native-windows/pull/5710#discussion_r470186502) [](ancestors = 470186502,470106454)